xline/src/storage/kvwatcher.rs

```diff
@@ -0,0 +1,312 @@
+            "WatchEvent receiver is closed"
+        );
+    }
+}
+
+/// KV watcher
+#[derive(Debug)]
+pub(crate) struct KvWatcher {
+    /// Inner data
+    inner: Arc<KvWatcherInner>,
+}
+
+/// KV watcher inner data
+#[derive(Debug)]
+pub(crate) struct KvWatcherInner {
+    /// KV storage
+    storage: Arc<KvStoreInner>,
+    /// watchers map
+    /// TODO: change to more effecient data structure
+    watcher_map: Mutex<HashMap<KeyRange, HashSet<Watcher>>>,
+}
+
+impl KvWatcher {
+    /// New `KvWatcher`
+    pub(crate) fn new(
+        storage: Arc<KvStoreInner>,
+        mut kv_update_rx: mpsc::Receiver<(i64, Vec<Event>)>,
+    ) -> Self {
+        let inner = Arc::new(KvWatcherInner::new(storage));
+        let inner_clone = Arc::clone(&inner);
+        let _handle = tokio::spawn(async move {
+            while let Some(updates) = kv_update_rx.recv().await {
+                inner_clone.handle_kv_updates(updates).await;
+            }
+        });
+        Self { inner }
+    }
+
+    /// Create a watch to KV store
+    pub(crate) async fn watch(
+        &self,
+        id: WatchId,
+        key_range: KeyRange,
+        start_rev: i64,
+        filters: Vec<i32>,
+        event_tx: mpsc::Sender<WatchEvent>,
+    ) -> (Watcher, Vec<Event>, i64) {
+        self.inner
+            .watch(id, key_range, start_rev, filters, event_tx)
+            .await
+    }
+
+    /// Cancel a watch from KV store
+    pub(crate) async fn cancel(&self, watcher: Watcher) -> i64 {
+        self.inner.cancel(watcher).await
+    }
+}
+
+impl KvWatcherInner {
+    /// New `KvWatchInner`
+    fn new(storage: Arc<KvStoreInner>) -> Self {
+        Self {
+            storage,
+            watcher_map: Mutex::new(HashMap::new()),
+        }
+    }
+
+    /// Create a watch to KV store
+    async fn watch(
+        &self,
+        id: WatchId,
+        key_range: KeyRange,
+        start_rev: i64,
+        filters: Vec<i32>,
+        event_tx: mpsc::Sender<WatchEvent>,
+    ) -> (Watcher, Vec<Event>, i64) {
+        let watcher = Watcher::new(key_range.clone(), id, start_rev, filters, event_tx);
+        let revision = self.storage.revision();
+        // TODO handle racing that new event is generated before watcher is registered
+        let events = if start_rev == 0 {
+            vec![]
+        } else {
+            self.storage
+                .get_event_from_revision(key_range.clone(), start_rev)
+        };
+
+        let _prev = self
+            .watcher_map
+            .lock()
+            .await
+            .entry(key_range)
+            .and_modify(|set| {
+                let _ = set.insert(watcher.clone());
+            })
+            .or_insert_with(|| {
+                let mut set = HashSet::new();
+                let _ = set.insert(watcher.clone());
+                set
+            });
+        (watcher, events, revision)
+    }
+
+    /// Cancel a watch from KV store
+    async fn cancel(&self, watcher: Watcher) -> i64 {
```

Comment:
We can pass in the watcher reference instead of value

Reply:
Done.